File: src/stripe_entitlements/checkout.py
from __future__ import annotations

import uuid
from dataclasses import dataclass
from datetime import datetime, timedelta
from typing import Protocol
from urllib.parse import urlsplit

import asyncpg


class CheckoutBusyError(RuntimeError):
    pass


class CheckoutCreationRejected(RuntimeError):
    """A deterministic pre-creation rejection for which releasing is safe."""


class CheckoutActiveSubscriptionError(RuntimeError):
    pass


@dataclass(frozen=True, slots=True)
class CheckoutReservation:
    account_id: str
    claim_token: str
    plan_key: str
    interval: str
    expires_at: datetime
    request_key: str
    session_id: str | None = None
    session_url: str | None = None


class CheckoutCreator(Protocol):
    async def create_checkout_session(
        self,
        *,
        account_id: str,
        customer_id: str | None,
        lookup_key: str,
        expected_currency: str,
        expected_unit_amount: int,
        expected_interval: str,
        claim_token: str,
        expires_at: datetime,
        customer_email: str | None,
        plan_key: str,
        interval: str,
    ) -> tuple[str, str]: ...


def _validate_visible(value: str, *, field: str, max_bytes: int) -> str:
    if (
        not isinstance(value, str)
        or not value
        or value != value.strip()
        or len(value.encode("utf-8")) > max_bytes
        or any(not character.isprintable() for character in value)
    ):
        raise ValueError(
            f"{field} must contain 1 to {max_bytes} visible UTF-8 bytes without padding"
        )
    return value


def _validate_session_identity(session_id: str, session_url: str) -> None:
    _validate_visible(session_id, field="Checkout Session id", max_bytes=255)
    _validate_visible(session_url, field="Checkout Session URL", max_bytes=2048)
    parsed = urlsplit(session_url)
    if (
        parsed.scheme != "https"
        or not parsed.netloc
        or parsed.username is not None
        or parsed.password is not None
    ):
        raise ValueError("Checkout Session URL must be an origin-safe HTTPS URL")


class CheckoutCoordinator:
    """A PostgreSQL-backed single-flight gate for Checkout Session creation."""

    def __init__(self, pool: asyncpg.Pool) -> None:
        self.pool = pool

    async def reserve(
        self,
        account_id: str,
        plan_key: str,
        interval: str,
        *,
        request_key: str | None = None,
        ttl: timedelta = timedelta(minutes=35),
        now: datetime | None = None,
    ) -> CheckoutReservation:
        if ttl <= timedelta(0):
            raise ValueError("Checkout claim TTL must be positive")
        if now is not None and now.tzinfo is None:
            raise ValueError("Checkout claim time must be timezone-aware")
        plan_key = _validate_visible(plan_key, field="plan_key", max_bytes=64)
        if "_" in plan_key:
            raise ValueError("plan_key cannot contain an underscore")
        if interval not in {"month", "year"}:
            raise ValueError("interval must be month or year")
        token = uuid.uuid4()
        request_key = (
            str(token)
            if request_key is None
            else _validate_visible(request_key, field="request_key", max_bytes=200)
        )
        async with self.pool.acquire() as conn, conn.transaction():
            effective_now = now or await conn.fetchval("select now()")
            assert isinstance(effective_now, datetime)
            expires_at = effective_now + ttl
            account = await conn.fetchrow(
                "select * from billing_accounts where id=$1::uuid for update", account_id
            )
            if account is None:
                raise KeyError("billing account not found")
            if account["stripe_subscription_id"] is not None or account["subscription_status"] in {
                "active",
                "past_due",
            }:
                raise CheckoutActiveSubscriptionError(
                    "an existing subscription must use the plan-change API"
                )
            existing = await conn.fetchrow(
                "select * from checkout_claims where account_id=$1::uuid for update",
                account_id,
            )
            if existing is not None and existing["expires_at"] > effective_now:
                if (
                    existing["client_request_key"] == request_key
                    and existing["plan_key"] == plan_key
                    and existing["plan_interval"] == interval
                ):
                    return CheckoutReservation(
                        account_id,
                        str(existing["claim_token"]),
                        plan_key,
                        interval,
                        existing["expires_at"],
                        request_key,
                        existing["session_id"],
                        existing["session_url"],
                    )
                raise CheckoutBusyError("an unexpired Checkout claim already exists")
            if existing is not None:
                await conn.execute(
                    "delete from checkout_claims where account_id=$1::uuid", account_id
                )
            await conn.execute(
                """insert into checkout_claims
                       (account_id,claim_token,plan_key,plan_interval,expires_at,
                        client_request_key)
                     values($1::uuid,$2,$3,$4,$5,$6)""",
                account_id,
                token,
                plan_key,
                interval,
                expires_at,
                request_key,
            )
        return CheckoutReservation(
            account_id, str(token), plan_key, interval, expires_at, request_key
        )

    async def attach_session(
        self, reservation: CheckoutReservation, session_id: str, session_url: str
    ) -> bool:
        _validate_session_identity(session_id, session_url)
        async with self.pool.acquire() as conn:
            updated = await conn.fetchval(
                """update checkout_claims set session_id=$3,session_url=$4
                     where account_id=$1::uuid and claim_token=$2::uuid
                     returning account_id""",
                reservation.account_id,
                reservation.claim_token,
                session_id,
                session_url,
            )
        return updated is not None

    async def release(self, reservation: CheckoutReservation) -> bool:
        async with self.pool.acquire() as conn:
            deleted = await conn.fetchval(
                """delete from checkout_claims
                     where account_id=$1::uuid and claim_token=$2::uuid returning account_id""",
                reservation.account_id,
                reservation.claim_token,
            )
        return deleted is not None

    async def completed_during_creation(self, reservation: CheckoutReservation) -> bool:
        """Detect an authorized webhook that consumed this claim before attach."""
        async with self.pool.acquire() as conn:
            row = await conn.fetchrow(
                """select a.stripe_subscription_id,c.claim_token
                     from billing_accounts a
                     left join checkout_claims c on c.account_id=a.id
                     where a.id=$1::uuid""",
                reservation.account_id,
            )
        return bool(
            row is not None and row["stripe_subscription_id"] and row["claim_token"] is None
        )

    async def create(
        self,
        creator: CheckoutCreator,
        *,
        account_id: str,
        customer_id: str | None,
        plan_key: str,
        interval: str,
        lookup_key: str,
        expected_currency: str,
        expected_unit_amount: int,
        expected_interval: str,
        request_key: str | None = None,
        customer_email: str | None = None,
    ) -> tuple[str, str]:
        reservation = await self.reserve(account_id, plan_key, interval, request_key=request_key)
        if reservation.session_id and reservation.session_url:
            return reservation.session_id, reservation.session_url
        try:
            session_id, url = await creator.create_checkout_session(
                account_id=account_id,
                customer_id=customer_id,
                lookup_key=lookup_key,
                expected_currency=expected_currency,
                expected_unit_amount=expected_unit_amount,
                expected_interval=expected_interval,
                claim_token=reservation.claim_token,
                expires_at=reservation.expires_at,
                customer_email=customer_email,
                plan_key=plan_key,
                interval=interval,
            )
        except CheckoutCreationRejected:
            await self.release(reservation)
            raise
        try:
            _validate_session_identity(session_id, url)
        except ValueError as exc:
            raise RuntimeError("Checkout creator returned an invalid Session identity") from exc
        # A generic Stripe/network exception has an unknown outcome. Keep this claim;
        # the caller must retry the same request key/claim token so Stripe idempotency
        # can return the original Session instead of opening a second payable Session.
        if not await self.attach_session(reservation, session_id, url):
            if await self.completed_during_creation(reservation):
                return session_id, url
            raise RuntimeError(
                "Checkout claim identity changed while Stripe was creating a session"
            )
        return session_id, url

File: typescript/src/checkout.ts
import { randomUUID } from "node:crypto";

import type { QueryResultRow } from "pg";

import type { Database } from "./database.js";
import type { TransactionClient } from "./db-types.js";
import {
  type CheckoutRequestSnapshot,
  StripeRequestSnapshotError,
  validateCheckoutRequestSnapshot,
} from "./stripe-request-snapshots.js";
import type { BillingInterval, PgTimestamp } from "./types.js";
import { isPlainRecord, requiredVisibleString } from "./validation.js";

export class CheckoutBusyError extends Error {}
export class CheckoutCreationRejected extends Error {}
export class CheckoutActiveSubscriptionError extends Error {}
export class CheckoutReplayUnsafeError extends Error {}

export interface CheckoutReservation {
  readonly accountId: string;
  readonly claimToken: string;
  readonly planKey: string;
  readonly interval: BillingInterval;
  readonly expiresAt: PgTimestamp;
  readonly expiresAtEpoch: bigint;
  readonly requestKey: string;
  readonly requestCustomerId?: string;
  readonly sessionId?: string;
  readonly sessionUrl?: string;
  readonly requestSnapshotVersion: number | null;
  readonly stripeRequestSnapshot: unknown;
}

export interface CheckoutCreator {
  prepareCheckoutSession(input: {
    readonly accountId: string;
    readonly customerId?: string;
    readonly lookupKey: string;
    readonly expectedCurrency: string;
    readonly expectedUnitAmount: bigint;
    readonly expectedInterval: BillingInterval;
    readonly claimToken: string;
    readonly expiresAtEpoch: bigint;
    readonly planKey: string;
    readonly interval: BillingInterval;
  }): Promise<unknown>;
  createCheckoutSessionFromSnapshot(
    snapshot: CheckoutRequestSnapshot,
  ): Promise<readonly [sessionId: string, sessionUrl: string]>;
}

interface AccountRow extends QueryResultRow {
  readonly stripe_customer_id: string | null;
  readonly stripe_subscription_id: string | null;
  readonly subscription_status: string;
}

interface ClockRow extends QueryResultRow {
  readonly effective_now: string;
  readonly expires_at: string;
  readonly expires_at_epoch: string;
}

interface ClaimRow extends QueryResultRow {
  readonly claim_token: string;
  readonly plan_key: string;
  readonly plan_interval: BillingInterval;
  readonly request_customer_id: string | null;
  readonly client_request_key: string | null;
  readonly session_id: string | null;
  readonly session_url: string | null;
  readonly active: boolean;
  readonly expires_at_text: string;
  readonly expires_at_epoch: string;
  readonly request_snapshot_version: number | null;
  readonly stripe_request_snapshot: unknown;
}

function reservationFromRow(
  accountId: string,
  requestKey: string,
  row: ClaimRow,
): CheckoutReservation {
  const base = {
    accountId,
    claimToken: row.claim_token,
    planKey: row.plan_key,
    interval: row.plan_interval,
    expiresAt: row.expires_at_text,
    expiresAtEpoch: BigInt(row.expires_at_epoch),
    requestKey,
    requestSnapshotVersion: row.request_snapshot_version,
    stripeRequestSnapshot: row.stripe_request_snapshot,
  } satisfies CheckoutReservation;
  return {
    ...base,
    ...(row.request_customer_id === null
      ? {}
      : { requestCustomerId: row.request_customer_id }),
    ...(row.session_id === null ? {} : { sessionId: row.session_id }),
    ...(row.session_url === null ? {} : { sessionUrl: row.session_url }),
  };
}

export function validateCheckoutSessionIdentity(
  sessionId: string,
  sessionUrl: string,
): void {
  requiredVisibleString(sessionId, "Checkout Session id", 255);
  requiredVisibleString(sessionUrl, "Checkout Session URL", 2048);
  let parsed: URL;
  try {
    parsed = new URL(sessionUrl);
  } catch {
    throw new TypeError(
      "Checkout Session URL must be an origin-safe HTTPS URL",
    );
  }
  if (
    parsed.protocol !== "https:" ||
    parsed.host.length === 0 ||
    parsed.username.length !== 0 ||
    parsed.password.length !== 0
  ) {
    throw new TypeError(
      "Checkout Session URL must be an origin-safe HTTPS URL",
    );
  }
}

async function databaseClock(
  tx: TransactionClient,
  ttlSeconds: number,
): Promise<ClockRow> {
  const result = await tx.query<ClockRow>(
    `with sampled as (select clock_timestamp() as effective_now),
          bounded as (
            select effective_now,
                   effective_now + make_interval(secs => $1::double precision) as expires_at
              from sampled
          )
     select to_char(effective_now at time zone 'UTC',
                    'YYYY-MM-DD"T"HH24:MI:SS.US"Z"') as effective_now,
            to_char(expires_at at time zone 'UTC',
                    'YYYY-MM-DD"T"HH24:MI:SS.US"Z"') as expires_at,
            floor(extract(epoch from expires_at))::bigint::text as expires_at_epoch
       from bounded`,
    [ttlSeconds],
  );
  const row = result.rows[0];
  if (row === undefined) {
    throw new Error("database clock query returned no row");
  }
  return row;
}

export class CheckoutCoordinator {
  readonly #database: Database;

  public constructor(database: Database) {
    this.#database = database;
  }

  public async reserve(
    accountId: string,
    planKeyInput: string,
    interval: BillingInterval,
    options: {
      readonly requestKey?: string;
      readonly ttlSeconds?: number;
    } = {},
  ): Promise<CheckoutReservation> {
    const ttlSeconds = options.ttlSeconds ?? 35 * 60;
    if (
      !Number.isSafeInteger(ttlSeconds) ||
      ttlSeconds <= 0 ||
      ttlSeconds > 86_400
    ) {
      throw new RangeError(
        "Checkout claim TTL must be a positive integer up to one day",
      );
    }
    const planKey = requiredVisibleString(planKeyInput, "plan_key", 64);
    if (planKey.includes("_")) {
      throw new TypeError("plan_key cannot contain an underscore");
    }
    if (interval !== "month" && interval !== "year") {
      throw new TypeError("interval must be month or year");
    }
    const token = randomUUID();
    const requestKey =
      options.requestKey === undefined
        ? token
        : requiredVisibleString(options.requestKey, "request_key", 200);

    return this.#database.transaction(async (tx) => {
      const accountResult = await tx.query<AccountRow>(
        "select * from billing_accounts where id=$1::uuid for update",
        [accountId],
      );
      const account = accountResult.rows[0];
      if (account === undefined) {
        throw new Error("billing account not found");
      }
      const clock = await databaseClock(tx, ttlSeconds);
      if (
        account.stripe_subscription_id !== null ||
        account.subscription_status === "active" ||
        account.subscription_status === "past_due"
      ) {
        throw new CheckoutActiveSubscriptionError(
          "an existing subscription must use the plan-change API",
        );
      }
      if (account.stripe_customer_id === null) {
        const pack = await tx.query<
          { readonly busy: boolean } & QueryResultRow
        >(
          `select exists(
             select 1 from credit_pack_orders
              where account_id=$1::uuid and payment_status='pending'
                and checkout_status <> 'expired'
                and claim_expires_at > $2::timestamptz
           ) as busy`,
          [accountId, clock.effective_now],
        );
        if (pack.rows[0]?.busy === true) {
          throw new CheckoutBusyError(
            "the first Stripe Customer Checkout is already in progress",
          );
        }
      }
      const existingResult = await tx.query<ClaimRow>(
        `select *, expires_at > $2::timestamptz as active,
                to_char(expires_at at time zone 'UTC',
                        'YYYY-MM-DD"T"HH24:MI:SS.US"Z"') as expires_at_text,
                floor(extract(epoch from expires_at))::bigint::text as expires_at_epoch
           from checkout_claims where account_id=$1::uuid for update`,
        [accountId, clock.effective_now],
      );
      const existing = existingResult.rows[0];
      if (existing !== undefined && existing.active) {
        if (
          existing.client_request_key === requestKey &&
          existing.plan_key === planKey &&
          existing.plan_interval === interval
        ) {
          return reservationFromRow(accountId, requestKey, existing);
        }
        throw new CheckoutBusyError(
          "an unexpired Checkout claim already exists",
        );
      }
      if (existing !== undefined) {
        await tx.query(
          "delete from checkout_claims where account_id=$1::uuid",
          [accountId],
        );
      }
      await tx.query(
        `insert into checkout_claims
           (account_id,claim_token,plan_key,plan_interval,expires_at,
            client_request_key,request_customer_id,request_snapshot_version)
         values($1::uuid,$2::uuid,$3,$4,$5::timestamptz,$6,$7,0)`,
        [
          accountId,
          token,
          planKey,
          interval,
          clock.expires_at,
          requestKey,
          account.stripe_customer_id,
        ],
      );
      return {
        accountId,
        claimToken: token,
        planKey,
        interval,
        expiresAt: clock.expires_at,
        expiresAtEpoch: BigInt(clock.expires_at_epoch),
        requestKey,
        requestSnapshotVersion: 0,
        stripeRequestSnapshot: null,
        ...(account.stripe_customer_id === null
          ? {}
          : { requestCustomerId: account.stripe_customer_id }),
      };
    });
  }

  public async freezeRequestSnapshot(
    reservation: CheckoutReservation,
    snapshot: CheckoutRequestSnapshot,
  ): Promise<unknown> {
    let result = await this.#database.query<
      {
        readonly request_snapshot_version: number | null;
        readonly stripe_request_snapshot: unknown;
      } & QueryResultRow
    >(
      `update checkout_claims set request_snapshot_version=1,
              stripe_request_snapshot=$3::jsonb
        where account_id=$1::uuid and claim_token=$2::uuid
          and request_snapshot_version=0 and stripe_request_snapshot is null
        returning request_snapshot_version,stripe_request_snapshot`,
      [reservation.accountId, reservation.claimToken, snapshot],
    );
    if (result.rows[0] === undefined) {
      result = await this.#database.query<
        {
          readonly request_snapshot_version: number | null;
          readonly stripe_request_snapshot: unknown;
        } & QueryResultRow
      >(
        `select request_snapshot_version,stripe_request_snapshot
           from checkout_claims
          where account_id=$1::uuid and claim_token=$2::uuid`,
        [reservation.accountId, reservation.claimToken],
      );
    }
    const frozen = result.rows[0];
    if (
      frozen?.request_snapshot_version !== 1 ||
      !isPlainRecord(frozen.stripe_request_snapshot)
    ) {
      throw new CheckoutReplayUnsafeError(
        "Checkout request snapshot could not be frozen safely",
      );
    }
    return frozen.stripe_request_snapshot;
  }

  public async attachSession(
    reservation: CheckoutReservation,
    sessionId: string,
    sessionUrl: string,
  ): Promise<boolean> {
    validateCheckoutSessionIdentity(sessionId, sessionUrl);
    const result = await this.#database.query<
      { readonly account_id: string } & QueryResultRow
    >(
      `update checkout_claims set session_id=$3,session_url=$4
         where account_id=$1::uuid and claim_token=$2::uuid
           and request_snapshot_version=1 and stripe_request_snapshot is not null
         returning account_id::text`,
      [reservation.accountId, reservation.claimToken, sessionId, sessionUrl],
    );
    return result.rowCount === 1;
  }

  public async release(reservation: CheckoutReservation): Promise<boolean> {
    const result = await this.#database.query<
      { readonly account_id: string } & QueryResultRow
    >(
      `delete from checkout_claims
         where account_id=$1::uuid and claim_token=$2::uuid
           and request_snapshot_version=0 and stripe_request_snapshot is null
           and session_id is null
         returning account_id::text`,
      [reservation.accountId, reservation.claimToken],
    );
    return result.rowCount === 1;
  }

  public async completedDuringCreation(
    reservation: CheckoutReservation,
  ): Promise<boolean> {
    const result = await this.#database.query<
      {
        readonly stripe_subscription_id: string | null;
        readonly claim_token: string | null;
      } & QueryResultRow
    >(
      `select a.stripe_subscription_id,c.claim_token::text
         from billing_accounts a
         left join checkout_claims c on c.account_id=a.id
        where a.id=$1::uuid`,
      [reservation.accountId],
    );
    const row = result.rows[0];
    return (
      row !== undefined &&
      row.stripe_subscription_id !== null &&
      row.claim_token === null
    );
  }

  #validatedFrozenSnapshot(
    reservation: CheckoutReservation,
  ): CheckoutRequestSnapshot {
    if (reservation.requestSnapshotVersion !== 1) {
      throw new CheckoutReplayUnsafeError(
        reservation.requestSnapshotVersion === null
          ? "this Checkout claim predates durable request snapshots; operator reconciliation is required"
          : "Checkout request snapshot state is invalid",
      );
    }
    try {
      return validateCheckoutRequestSnapshot(
        reservation.stripeRequestSnapshot,
        {
          kind: "subscription",
          accountId: reservation.accountId,
          requestIdentity: reservation.claimToken,
          interval: reservation.interval,
          offeringKey: reservation.planKey,
          expiresAt: reservation.expiresAtEpoch,
          customerId: reservation.requestCustomerId ?? null,
        },
      );
    } catch (error: unknown) {
      if (error instanceof StripeRequestSnapshotError) {
        throw new CheckoutReplayUnsafeError(
          "the persisted Checkout request snapshot is invalid; operator reconciliation is required",
          { cause: error },
        );
      }
      throw error;
    }
  }

  async #executeFrozen(
    creator: CheckoutCreator,
    reservation: CheckoutReservation,
    snapshot: CheckoutRequestSnapshot,
  ): Promise<readonly [sessionId: string, sessionUrl: string]> {
    if (
      reservation.sessionId !== undefined &&
      reservation.sessionUrl !== undefined
    ) {
      return [reservation.sessionId, reservation.sessionUrl];
    }
    const created = await creator.createCheckoutSessionFromSnapshot(snapshot);
    const [sessionId, sessionUrl] = created;
    try {
      validateCheckoutSessionIdentity(sessionId, sessionUrl);
    } catch (error: unknown) {
      throw new Error("Checkout creator returned an invalid Session identity", {
        cause: error,
      });
    }
    if (!(await this.attachSession(reservation, sessionId, sessionUrl))) {
      if (await this.completedDuringCreation(reservation)) {
        return created;
      }
      throw new Error(
        "Checkout claim identity changed while Stripe was creating a session",
      );
    }
    return created;
  }

  public async recoverFrozen(
    creator: CheckoutCreator,
    input: {
      readonly accountId: string;
      readonly planKey: string;
      readonly interval: BillingInterval;
      readonly requestKey: string;
    },
  ): Promise<readonly [sessionId: string, sessionUrl: string] | undefined> {
    const planKey = requiredVisibleString(input.planKey, "plan_key", 64);
    const requestKey = requiredVisibleString(
      input.requestKey,
      "request_key",
      200,
    );
    if (
      planKey.includes("_") ||
      (input.interval !== "month" && input.interval !== "year")
    ) {
      throw new TypeError("invalid frozen Checkout recovery identity");
    }
    const result = await this.#database.query<ClaimRow>(
      `select *,true as active,
              to_char(expires_at at time zone 'UTC',
                      'YYYY-MM-DD"T"HH24:MI:SS.US"Z"') as expires_at_text,
              floor(extract(epoch from expires_at))::bigint::text as expires_at_epoch
         from checkout_claims
        where account_id=$1::uuid and client_request_key=$2
          and plan_key=$3 and plan_interval=$4
          and expires_at>clock_timestamp()
          and request_snapshot_version=1 and stripe_request_snapshot is not null`,
      [input.accountId, requestKey, planKey, input.interval],
    );
    const row = result.rows[0];
    if (row === undefined) {
      return undefined;
    }
    const reservation = reservationFromRow(input.accountId, requestKey, row);
    return this.#executeFrozen(
      creator,
      reservation,
      this.#validatedFrozenSnapshot(reservation),
    );
  }

  public async create(
    creator: CheckoutCreator,
    input: {
      readonly accountId: string;
      readonly planKey: string;
      readonly interval: BillingInterval;
      readonly lookupKey: string;
      readonly expectedCurrency: string;
      readonly expectedUnitAmount: bigint;
      readonly expectedInterval: BillingInterval;
      readonly requestKey?: string;
    },
  ): Promise<readonly [sessionId: string, sessionUrl: string]> {
    if (input.interval !== input.expectedInterval) {
      throw new CheckoutCreationRejected(
        "Checkout interval does not match the catalog expectation",
      );
    }
    const reservation = await this.reserve(
      input.accountId,
      input.planKey,
      input.interval,
      {
        ...(input.requestKey === undefined
          ? {}
          : { requestKey: input.requestKey }),
      },
    );
    let snapshot: CheckoutRequestSnapshot;
    if (reservation.requestSnapshotVersion === null) {
      throw new CheckoutReplayUnsafeError(
        "this Checkout claim predates durable request snapshots; operator reconciliation is required",
      );
    }
    if (reservation.requestSnapshotVersion === 1) {
      snapshot = this.#validatedFrozenSnapshot(reservation);
    } else if (
      reservation.requestSnapshotVersion !== 0 ||
      reservation.stripeRequestSnapshot !== null
    ) {
      throw new CheckoutReplayUnsafeError(
        "Checkout request snapshot state is invalid",
      );
    } else {
      let prepared: CheckoutRequestSnapshot;
      try {
        prepared = validateCheckoutRequestSnapshot(
          await creator.prepareCheckoutSession({
            accountId: reservation.accountId,
            ...(reservation.requestCustomerId === undefined
              ? {}
              : { customerId: reservation.requestCustomerId }),
            lookupKey: input.lookupKey,
            expectedCurrency: input.expectedCurrency,
            expectedUnitAmount: input.expectedUnitAmount,
            expectedInterval: input.expectedInterval,
            claimToken: reservation.claimToken,
            expiresAtEpoch: reservation.expiresAtEpoch,
            planKey: reservation.planKey,
            interval: reservation.interval,
          }),
          {
            kind: "subscription",
            accountId: reservation.accountId,
            requestIdentity: reservation.claimToken,
            lookupKey: input.lookupKey,
            currency: input.expectedCurrency,
            unitAmount: input.expectedUnitAmount,
            interval: input.expectedInterval,
            offeringKey: reservation.planKey,
            expiresAt: reservation.expiresAtEpoch,
            customerId: reservation.requestCustomerId ?? null,
          },
        );
      } catch (error: unknown) {
        if (error instanceof CheckoutCreationRejected) {
          await this.release(reservation);
        }
        throw error;
      }
      const frozen = await this.freezeRequestSnapshot(reservation, prepared);
      snapshot = this.#validatedFrozenSnapshot({
        ...reservation,
        requestSnapshotVersion: 1,
        stripeRequestSnapshot: frozen,
      });
    }
    return this.#executeFrozen(creator, reservation, snapshot);
  }
}
